README.md
# Redmine Serial Number Field

[English »](https://github.com/matsukei/redmine_serial_number_field/blob/master/README.en.md)

Redmine に自動で連番を付加するカスタムフィールドを提供するプラグインです。

## Features

* チケットに対するカスタムフィールドの書式として「自動採番」が利用可能になります。
  * カスタムフィールドの新規作成後は、正規表現を編集することは、できません。
  * チケットが登録できる、全てのユーザーは、自動採番の権限があります。
* 指定した正規表現に基づいて、チケット登録や更新時（一括操作も含む）に自動的に採番します。
  * カスタムフィールド単位での採番となります。同じカスタムフィールドを複数のプロジェクトで使用している場合は、それらのプロジェクトで連続した採番になります。
  * カスタムフィールド項目は、チケットの閲覧時には表示します。しかし、登録および更新時には表示しません。
* チケットのフィルタ条件、検索対象など、カスタムフィールドの基本的なオプションも利用可能です。

### Notes

#### 採番済みチケットのトラッカーやプロジェクトを変更した場合

* 変更後のトラッカーが同じカスタムフィールドを持っていない場合、採番された連番は削除されます。
* 変更後のトラッカーが同じカスタムフィールドを持っている場合、採番された連番は変化しません。

#### ワークフローのカスタムフィールドに対する権限を設定した場合

* 自動採番のカスタムフィールドを読み取り専用に設定すると動作がおかしくなる場合があります。

## Usage

1. チケットの新しいカスタムフィールドを作成します。
2. 項目「書式」を自動採番に変更します。
3. 項目「正規表現」に年表記書式や連番書式を指定します。
4. フィルタや検索条件として使いたい場合は、適宜チェックを入れます。
5. 自動採番をしたいトラッカーとプロジェクトを指定します。
6. 完了です。
    * 新たにチケットを作成すれば自動で採番されます。
    * 既に作成されたチケットは何かしら更新すれば採番されます。

## Screenshot

*Administration > Custom fields > Issues > 自動採番*

![usage.png](https://github.com/matsukei/redmine_serial_number_field/blob/master/doc/images/usage.png)

*Issues*

![issues.png](https://github.com/matsukei/redmine_serial_number_field/blob/master/doc/images/issues.png)

## Supported versions

* Redmine 4.1 (Ruby 2.6)

## Format specifications

|採番対象の日付             |年表記書式 |年度(4/1 - 3/31)       |例: 正規表現             |例: 結果(2015-03-31)   |
|---------------------------|-----------|-----------------------|-------------------------|-----------------------|
|Issue#created_on           |`yyyy`     |No                     |`{yyyy}-{0000}`          |`2015-0001`            |
|^                          |`yy`       |No                     |`{yy}-{0000}`            |`15-0001`              |
|^                          |`YYYY`     |Yes                    |`{YYYY}-{0000}`          |`2014-0001`            |
|^                          |`YY`       |Yes                    |`{YY}-{0000}`            |`14-0001`              |
|^                          |`ISO`      |No                     |`{ISO}-{0000}`           |`20150331-0001`        |

* OK
  * `{000000}` #=> `000001`
  * `ABC-{yy}-{00}` #=> `ABC-15-01`
* NG
  * 末尾が連番書式でない場合
    * e.g. `ABC-{000}-{yy}`
  * 年表記書式、連番書式でない場合
    * e.g. `{abc}-{yy}-{000}`

## Install

1. `your_redmine_path/plugins/redmine_serial_number_field/` に clone もしくはダウンロードしたソースを配置します
2. `$ cd your_redmine_path/`
3. `$ bundle install`
4. Redmineを再起動してください

## License

[The MIT License](https://opensource.org/licenses/MIT)


CHANGELOG.md
## 3.0.0

- Added ISO format #8 (@rambowm)
- Dropped Redmine 4.0 support
- Redmine 4.1 support
- Fixed automatic serial number when copying issue
  - チケットコピー時に自動採番されない #6 (@subaru019)
  - [Review comments for this plugin on redmine.org](https://www.redmine.org/plugins/redmine_serial_number_field) (Jorge Garcia, Quan VN, qn wang)

## 2.0.0

- Support Redmine 4.0 (@nanego)
- Support French locale (@nanego)
- Drop support Redmine version 3.4 lower (@maeda-m)

## 1.0.0

- Changed the range to count sequential numbers from project unit to whole sharing as custom field unit #4 (@neocosmic2)

## 0.9.0

- Support Turkish locale #3 (@atopcu)
- Allow only sequence numbers #2 (@y-arai)
- Drop support Redmine versions 2.6 and 3.1 (@maeda-m)


lib/format.rb
module SerialNumberField
  class Format < Redmine::FieldFormat::Base
    NAME = 'serial_number'

    add NAME
    self.searchable_supported = true
    self.customized_class_names = %w(Issue)
    self.form_partial = 'custom_fields/formats/serial_number'

    FORMAT_WRAPPER = /\{(.+?)\}/
    DATE_FORMATS = {
      :'yyyy' => { :strftime => '%Y',     :financial_year => false },
      :'yy'   => { :strftime => '%y',     :financial_year => false },
      :'YYYY' => { :strftime => '%Y',     :financial_year => true },
      :'YY'   => { :strftime => '%y',     :financial_year => true },
      :'ISO'  => { :strftime => '%Y%m%d', :financial_year => false }
    }

    def validate_custom_field(custom_field)
      value = custom_field.regexp
      errors = []

      errors << [:regexp, :end_must_numeric_format_in_serial_number] unless value =~ /\{0+\}\Z/
      replace_format_value(custom_field) do |format_value|
        unless format_value =~ /\A0+\Z/
          errors << [:regexp, :invalid_format_in_serial_number] unless date_format_keys.include?(format_value)
        end
      end

      errors.uniq
    end

    def generate_value(custom_field, issue)
      datetime = issue.created_on.to_datetime || DateTime.now
      value = max_custom_value(custom_field, datetime)

      if value.present?
        value.next
      else
        generate_first_value(custom_field, datetime)
      end
    end

    private

      def date_format_keys
        DATE_FORMATS.stringify_keys.keys
      end

      def max_custom_value(custom_field, datetime)
        matcher = generate_matcher(custom_field, datetime)
        custom_values = custom_field.custom_values.map(&:value)
        # custom_values #=> e.g. ['2014-001', '2014-002', '14-0001', ...]
        custom_values.select { |value| value =~ matcher }.sort.last
      end

      def replace_format_value(custom_field)
        if block_given?
          custom_field.regexp.gsub(FORMAT_WRAPPER) do |format_value_with_brace|
            # format_value_with_brace #=> e.g. '{yy}', '{0000}'
            # $1.clone #=> e.g. 'yy', '0000'
            yield($1.clone)
          end
        end
      end

      # e.g. /2015-\d{3}/
      def generate_matcher(custom_field, datetime)
        matcher_str = replace_format_value(custom_field) do |format_value|
          if date_format_keys.include?(format_value)
            generate_date_value(format_value, datetime)
          else
            # TODO:
            #
            # CustomField of A.regexp #=> '{yy}-0{000}'
            #   matcher of A: /15-0\d{3}/
            # CustomField of B.regexp #=> '{yy}-{0000}'
            #   matcher of B: /15-\d{4}/
            #
            # if CustomValue is '15-0999',
            # CustomValue matches matcher of both A and B
            generate_number_matcher(format_value)
          end
        end

        Regexp.new(matcher_str)
      end

      # e.g. '\d{3}'
      def generate_number_matcher(format_value)
        ['\d', '{', format_value.size.to_s, '}'].join
      end

      def generate_first_value(custom_field, datetime)
        replace_format_value(custom_field) do |format_value|
          if date_format_keys.include?(format_value)
            generate_date_value(format_value, datetime)
          else
            generate_number_value(format_value)
          end
        end
      end

      # e.g. '2015'
      def generate_date_value(format_value, datetime)
        parse_conf = DATE_FORMATS[format_value.to_sym]
        if parse_conf.key?(:financial_year) && parse_conf[:financial_year]
          # TODO initializers
          DateTime.fiscal_zone = :japan
          datetime = datetime.beginning_of_financial_year
        end

        datetime.strftime(parse_conf[:strftime])
      end

      # e.g. '0001'
      def generate_number_value(format_value)
        '1'.rjust(format_value.size, '0')
      end

  end
end
